artifacthub/library/general/storageclass/1.1.0/template.yaml

```diff
@@ -0,0 +1,155 @@
+          obj.apiVersion == "v1"
+          obj.kind == "PersistentVolumeClaim"
+        }
+
+        is_statefulset(obj) {
+          obj.apiVersion == "apps/v1"
+          obj.kind == "StatefulSet"
+        }
+
+        violation[{"msg": msg}] {
+          not data.inventory.cluster["storage.k8s.io/v1"]["StorageClass"]
+          msg := sprintf("StorageClasses not synced. Gatekeeper may be misconfigured. Please have a cluster-admin consult the documentation.", [])
+        }
+
+        storageclass_allowed(name) {
+          data.inventory.cluster["storage.k8s.io/v1"]["StorageClass"][name]
+          # support both direct use of * and as the default value
+          object.get(input.parameters, "allowedStorageClasses", ["*"])[_] == "*"
+        }
+
+        storageclass_allowed(name) {
+          data.inventory.cluster["storage.k8s.io/v1"]["StorageClass"][name]
+          input.parameters.allowedStorageClasses[_] == name
+        }
+
+        violation[{"msg": pvc_storageclass_badname_msg}] {
+          is_pvc(input.review.object)
+          not storageclass_allowed(input.review.object.spec.storageClassName)
+        }
+        pvc_storageclass_badname_msg := sprintf("pvc did not specify a valid storage class name <%v>. Must be one of [%v]", args) {
+          input.parameters.includeStorageClassesInMessage
+          object.get(input.parameters, "allowedStorageClasses", null) == null
+          args := [
+            input.review.object.spec.storageClassName,
+            concat(", ", [n | data.inventory.cluster["storage.k8s.io/v1"]["StorageClass"][n]])
+          ]
+        } else := sprintf("pvc did not specify an allowed and valid storage class name <%v>. Must be one of [%v]", args) {
+          input.parameters.includeStorageClassesInMessage
+          object.get(input.parameters, "allowedStorageClasses", null) != null
+          sc := {n | data.inventory.cluster["storage.k8s.io/v1"]["StorageClass"][n]} & {x | x = object.get(input.parameters, "allowedStorageClasses", [])[_]}
+          args := [
+            input.review.object.spec.storageClassName,
+            concat(", ", sc)
+          ]
+        } else := sprintf(
+          "pvc did not specify a valid storage class name <%v>.",
+          [input.review.object.spec.storageClassName]
+        )
+
+        violation[{"msg": pvc_storageclass_noname_msg}] {
+          is_pvc(input.review.object)
+          not input.review.object.spec.storageClassName
+        }
+        pvc_storageclass_noname_msg := sprintf("pvc did not specify a storage class name. Must be one of [%v]", args) {
+          input.parameters.includeStorageClassesInMessage
+          args := [
+            concat(", ", [n | data.inventory.cluster["storage.k8s.io/v1"]["StorageClass"][n]])
+          ]
+        } else := sprintf(
+          "pvc did not specify a storage class name.",
+          []
+        )
+
+        violation[{"msg": statefulset_vct_badname_msg(vct)}] {
+          is_statefulset(input.review.object)
```

Comment:
Why statefulset vs. pods?

Reply:
Hey @maxsmythe - This was the original design/code (#214), and my (initial) goal was to minimize changes to ensure backwards compatibility.

This CT also applies to `PersistentVolumeClaim`, so it would cover pods from that angle.  My assumption is `StatefulSet` was included to catch earlier in the stack/block on admission.

Reply:
Ah, it looks like stateful sets specifically have volume claim templates.